src/main/java/dev/binclub/javaception/classfile/ClassFileParser.java

```diff
@@ -1,4 +1,235 @@
+	DataInputStream dis;
+	Object[] constantPool;
+	ClassInfoCP[] interfaces;
+	FieldInfo[] fields;
+	MethodInfo[] methods;
+	AttributeInfo[] attributes;
+
+	public ClassFileParser(byte[] classBytes) throws Throwable {
+		if (classBytes.length < 4) {
+			// need atleast 4 bytes get magic number
+			throw new Throwable("Invalid class file");
+		}
+
+		parseClass(classBytes);
+
+	}
+
+	public void parseClass(byte[] classBytes) throws Throwable {
+
+		dis = new DataInputStream(new ByteArrayInputStream(classBytes));
+		int magic = dis.readInt();
+		if (magic != 0xCAFEBABE) {
+			throw new Throwable("Invalid Class File");
+		}
+		minorVersion = dis.readUnsignedShort();
+		majorVersion = dis.readUnsignedShort();
+		constantPoolCount = dis.readUnsignedShort();
+		constantPool = new Object[constantPoolCount - 1];
+		// force enum to resolve
+		System.out.println(ConstantTypes.values().length);
+		parseConstantPool();
+		access = dis.readUnsignedShort();
+		int classNameIndex = dis.readUnsignedShort();
+		className = ((ClassInfo) constantPool[classNameIndex - 1]).getClassName(constantPool);
+		System.out.println("ClassName " + className);
+		int superClassNameIndex = dis.readUnsignedShort();
+		superClass = ((ClassInfo) constantPool[superClassNameIndex - 1]).getClassName(constantPool);
+		System.out.println("SuperClass " + superClass);
+		interfacesCount = dis.readUnsignedShort();
+		getInterfaces();
+		fieldsCount = dis.readUnsignedShort();
+		readFields();
+		methodsCount = dis.readUnsignedShort();
+		readMethods();
+		attributesCount = dis.readUnsignedShort();
+		if (attributesCount != 0) {
+			attributes = new AttributeInfo[attributesCount];
+			for (int i = 0; i < attributesCount; i++) {
+				attributes[i] = readAttribute(dis, constantPool);
+			}
+		}
+	}
+
+	public void readMethods() throws IOException {
+		methods = new MethodInfo[methodsCount];
+		for (int i = 0; i < methodsCount; i++) {
+			int access = dis.readUnsignedShort();
+			int nameIndex = dis.readUnsignedShort();
+			int descriptorIndex = dis.readUnsignedShort();
+			int attributesCount = dis.readUnsignedShort();
+			AttributeInfo[] attributes = null;
+			if (attributesCount != 0) {
+				attributes = new AttributeInfo[attributesCount];
+				for (int j = 0; j < attributesCount; j++) {
+					attributes[j] = readAttribute(dis, constantPool);
+				}
+			}
+			methods[i] = new MethodInfo(access, nameIndex, descriptorIndex, attributes);
+
+		}
+	}
+
+	public void readFields() throws Throwable {
+		fields = new FieldInfo[fieldsCount];
+		for (int i = 0; i < fieldsCount; i++) {
+			int access = dis.readUnsignedShort();
+			int nameIndex = dis.readUnsignedShort();
+			int descriptorIndex = dis.readUnsignedShort();
+			int attributesCount = dis.readUnsignedShort();
+			AttributeInfo[] attributes = null;
+			if (attributesCount != 0) {
+				attributes = new AttributeInfo[attributesCount];
+				for (int j = 0; j < attributesCount; j++) {
+					attributes[j] = readAttribute(dis, constantPool);
```

Comment:
`readAttribute` can sometimes return null, meaning there will be unused array slots of null values, perhaps it would be more efficient to use an `ArrayList` and add attributes to it only if they are non null?